v1.1/parallel-statement-execution.md

```diff
@@ -0,0 +1,128 @@
+
+Consider a scenario where a user's last name, favorite movie, and favorite song are to be updated on a social networking application. Suppose the database has three tables: username, movies, and songs, that need to be updated to update the user information. Also suppose that the user's ID is 1. Then the traditional transaction to update the user's information is as follows:
+
+~~~ sql
+> BEGIN;
+> UPDATE username SET last_name = 'Smith' WHERE id = 1;
+> UPDATE movies SET favorite_movie = 'The Matrix' WHERE id = 1;
+> UPDATE songs SET favorite_song = 'All this time' WHERE id = 1;
+> COMMIT;
+~~~
+
+In this case, the statements are executed sequentially. Sequential execution of SQL statements returns a return value immediately from the server. After the return value is received, the next SQL statement is executed. The following conceptual diagram shows the sequential execution of the statements.
+
+<img src="{{ 'images/Sequential_Statement_Execution.png' | relative_url }}" alt="CockroachDB Parallel Statement Execution Error Mismatch" style="border:1px solid #eee;max-width:100%" />
+
+Traditional SQL engines execute the statements sequentially without affecting performance. But in case of globally-distributed CockroachDB clusters, sequential execution of statements adds to aggregate latency. This is because to maintain strong consistency across a globally-distributed cluster, CockroachDB replicates writes using the Raft [consensus protocol](https://www.cockroachlabs.com/blog/consensus-made-thrive/). The protocol requires majority of replicas to agree before a write is committed. Replicas always live on separate nodes, and communication between these nodes takes time. Executing SQL statements sequentially leads to higher cumulative communication latency. 
+
+With parallel statement execution, however, multiple SQL statements within a transaction are executed at the same time, thereby reducing the aggregate latency over multiple SQL statements. 
+
+## How Parallel Statement Execution Works
+
+SQL statements within a single transaction can be executed in parallel if the statements are independent. CockroachDB considers SQL statements within a single transaction to be independent if their execution can be safely reordered without affecting their results. 
+
+As seen earlier, sequential execution of SQL statements returns a return value immediately from the server. The next statement is executed only after the return value is received. Appending the `RETURNING NOTHING` clause with SQL statements prevents the server from sending return values and allows CockroachDB to execute the statements in parallel. To execute statements in parallel, append the `RETURNING NOTHING` suffix to the SQL statements. The statements are executed in parallel till CockroachDB encounters a **barrier statement**. A barrier statement is a statement that needs to be executed sequentially.
```

Comment:
"prevents" might have the wrong connotation. Try "stops". I think it's also important to note here somewhere that we return to the client before executing the statement. This is actually the key that allows us to execute multiple statements in parallel.

Reply:
Also, s/till/until/

Reply:
> A barrier statement is a statement that needs to be executed sequentially.

This is going to be confusing when people read it first. I'd just say "A barrier statement is any other statement without the `RETURNING NOTHING` suffix"

Reply:
About noting that we return to client: I think it's a technical detail that is already evident from the diagram. Spelling it out might distract from the main point, which is we stop server from sending return values, thereby allowing parallel execution. Thoughts?

Reply:
I think this is the main point though. The client can't send the next statement until it hears something from the server (that's why the conversation analogy is so good!). I would change `the client sends` to `the client can then send` here and below. The reason for this is subtle but kind of important. The reader of these docs is the client. We're describing the contract we (the server) are making with their application (the client).

> sequential execution of SQL statements returns a return value immediately from the server

It's not immediate, it's after executing the statement, right?

Reply:
I reworded the section a bit:
As seen earlier, sequential execution of SQL statements returns a return value after executing the statement. The client can send the next statement to be executed only after it receives the return value from the server. In CockroachDB, appending the `RETURNING NOTHING` clause with SQL statements causes the server to send an acknowledgement immediately instead of completing the execution and then sending the return value to the client. The client can then send the next statement to be executed on receiving the acknowledgement. This allows CockroachDB to execute the statements in parallel. The statements are executed in parallel until CockroachDB encounters a **barrier statement**. A barrier statement is any statement without the `RETURNING NOTHING` suffix and is executed sequentially.

Does that seem clearer?